drivers/kscan/kscan_matrix_poll.c

```diff
@@ -0,0 +1,323 @@
+	/* Track previous "elapsed clock cycles" per matrix scan. This
+	 * is used to calculate the debouncing time for every key
+	 */
+	uint8_t scan_clk_cycle[SCAN_OCURRENCES];
+	uint8_t scan_cycles_idx;
+	struct k_thread thread;
+	struct kscan_matrix_config config;
+
+	K_KERNEL_STACK_MEMBER(thread_stack, TASK_STACK_SIZE);
+};
+static struct kscan_matrix_data kb_poll;
+
+#define KSCAN_DEV_NAME DT_LABEL(DT_ALIAS(kscan0))
+
+static const struct device *kscan_dev;
+struct k_sem poll_lock;
+
+static bool is_matrix_ghosting(const uint8_t *state)
+{
+	/* matrix keyboard designs are suceptible to ghosting.
+	 * An extra key appears to be pressed when 3 keys
+	 * belonging to the same block are pressed.
+	 * for example, in the following block
+	 *
+	 * . . w . q .
+	 * . . . . . .
+	 * . . . . . .
+	 * . . m . a .
+	 *
+	 * the key m would look as pressed if the user pressed keys
+	 * w, q and a simultaneously. A block can also be formed,
+	 * with not adjacent columns.
+	 */
+	for (int c = 0; c < CONFIG_KSCAN_MATRIX_MAX_COLUMNS; c++) {
+		if (!state[c]) {
+			continue;
+		}
+
+		for (int c_n = c + 1; c_n < CONFIG_KSCAN_MATRIX_MAX_COLUMNS; c_n++) {
+			/* we and the columns to detect a "block".
+			 * this is an indication of ghosting, due to current
+			 * flowing from a key which was never pressed. in our
+			 * case, current flowing is a bit set to 1 as we
+			 * flipped the bits when the matrix was scanned.
+			 * now we or the colums using z&(z-1) which is
+			 * non-zero only if z has more than one bit set.
+			 */
+			uint8_t common_row_bits = state[c] & state[c_n];
+
+			if (common_row_bits & (common_row_bits - 1)) {
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+static bool read_keyboard_matrix(uint8_t *new_state)
+{
+	int row;
+	uint8_t key_event = 0U;
+
+	for (int col = 0; col < CONFIG_KSCAN_MATRIX_MAX_COLUMNS; col++) {
+		kscan_drive_matrix_column(kscan_dev, col);
+
+		/* Allow the matrix to stabilize before reading it */
+		k_msleep(10);
+
+		kscan_read_matrix_row(kscan_dev, &row);
+		new_state[col] = row & 0xFF;
+		key_event |= row;
+	}
+
+	kscan_drive_matrix_column(kscan_dev, KEYBOARD_COLUMN_DRIVE_NONE);
+
+	return key_event != 0U ? true : false;
+}
+
+static bool check_key_events(const struct device *dev)
+{
+	uint8_t matrix_new_state[CONFIG_KSCAN_MATRIX_MAX_COLUMNS] = { 0U };
+	bool key_pressed = false;
+	uint32_t cycles_now = k_cycle_get_32();
+
+	if (++kb_poll.scan_cycles_idx >= SCAN_OCURRENCES) {
+		kb_poll.scan_cycles_idx = 0U;
+	}
+
+	kb_poll.scan_clk_cycle[kb_poll.scan_cycles_idx] = cycles_now;
+
+	/* Scan the matrix */
+	key_pressed = read_keyboard_matrix(matrix_new_state);
+
+	for (int c = 0; c < CONFIG_KSCAN_MATRIX_MAX_COLUMNS; c++) {
+		LOG_DBG("U%x, P%x, N%x", kb_poll.matrix_unstable_state[c],
+			kb_poll.matrix_previous_state[c], matrix_new_state[c]);
+	}
+
+	/* Abort if ghosting is detected */
+	if (is_matrix_ghosting(matrix_new_state)) {
+		return false;
```

Comment:
The code of the kscan matrix poll task is extracted from the [kscan_mchp_xec](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/kscan/kscan_mchp_xec.c#L249) driver and it return false.
I don't understand why it return false here.
Anyway, I have changed the task to return  any_pressed to meet the  logic of Chromium EC keyboard_scan.c because the embedded keyboard scan drivers are majorly used in the Chromebook application.